library/src/blas2/rocblas_hemv.hpp

```diff
@@ -0,0 +1,216 @@
+
+    for(col = ty; col < n; col += DIM_Y)
+    {
+        if(ind < n)
+        {
+            if(col > ind)
+                tmp_A = upper ? A[ind + col * lda] : conj(A[col + ind * lda]);
+            else if(col < ind)
+                tmp_A = !upper ? A[ind + col * lda] : conj(A[col + ind * lda]);
+            else if(col == ind)
+                tmp_A = std::real(A[ind + col * lda]);
+
+            res_A += tmp_A * x[(col)*incx];
+        }
+    }
+
+    sdata[tx + ty * DIM_X] = res_A;
+    __syncthreads();
+
+    ind = hipBlockIdx_x * DIM_X + thread_id;
+    if(thread_id < DIM_X)
+    {
+        for(rocblas_int i = 1; i < DIM_Y; i++)
+            sdata[thread_id] += sdata[thread_id + DIM_X * i];
+
+        if(ind < n)
+        {
+            if(beta != 0)
+            {
+                y[ind * incy] = (alpha * sdata[thread_id]) + (beta * y[ind * incy]);
+            }
+            else
+            {
+                y[ind * incy] = alpha * sdata[thread_id];
+            }
+        }
+    }
+}
+
+/**
+  *  U is either: const T* OR T
+  *  V is either: const T* OR const T* const*
+  *  W is either:       T* OR       T* const*
+  */
+template <rocblas_int DIM_X, rocblas_int DIM_Y, typename U, typename V, typename W>
+__global__ void hemvn_kernel(rocblas_fill   uplo,
+                             rocblas_int    n,
+                             U              alpha_device_host,
+                             V              Aa,
+                             ptrdiff_t      shifta,
+                             rocblas_int    lda,
+                             rocblas_stride strideA,
+                             V              xa,
+                             ptrdiff_t      shiftx,
+                             rocblas_int    incx,
+                             rocblas_stride stridex,
+                             U              beta_device_host,
+                             W              ya,
+                             ptrdiff_t      shifty,
+                             rocblas_int    incy,
+                             rocblas_stride stridey)
+{
+    rocblas_int num_threads = hipBlockDim_x * hipBlockDim_y * hipBlockDim_z;
+    if(DIM_X * DIM_Y != num_threads)
+        return; // need to launch exactly the same number of threads as template parameters indicate
+
+    const auto* A = load_ptr_batch(Aa, hipBlockIdx_y, shifta, strideA);
+    const auto* x = load_ptr_batch(xa, hipBlockIdx_y, shiftx, stridex);
+    auto*       y = load_ptr_batch(ya, hipBlockIdx_y, shifty, stridey);
+
+    auto alpha = load_scalar(alpha_device_host);
+    auto beta  = load_scalar(beta_device_host);
+
+    hemvn_kernel_calc<DIM_X, DIM_Y>(uplo, n, alpha, A, lda, x, incx, beta, y, incy);
+}
+
+/**
+  *  U is always: const T* (either host or device)
+  *  V is either: const T* OR const T* const*
+  *  W is either:       T* OR       T* const*
+  */
+template <typename U, typename V, typename W>
+rocblas_status rocblas_hemv_template(rocblas_handle handle,
```

Comment:
To me we should not name this function rocblas_hemv_template.
rocblas_hemv_template should not have strides and batch_count parameters.

rocblas_hemv_template(rocblas_fill   uplo,
                                     rocblas_int    n,
                                     U              alpha,
                                     V              A,
                                     rocblas_int    offseta,
                                     rocblas_int    lda,
                                     V              x,
                                     rocblas_int    offsetx,
                                     rocblas_int    incx,
                                     U              beta,
                                     W              y,
                                     rocblas_int    offsety,
                                     rocblas_int    incy)

Reply:
I think we'll have to open a thread about this to get input from everyone (in the new year, likely). I thought that the group had decided on this way awhile back but I could very well be mistaken.
To me, it seems that there are two options/opinions:

A) Do what I've done here, have one general rocblas_hemv_template which is called from each .cpp file, which can handle all inputs (batch_count, strides, offsets, etc.). This is currently used in the functions:
    (1) copy
    (2) rot
    (3) rotg
    (4) rotm
    (5) rotmg
    (6) scal
    (7) swap
    (8) gemv
    (9) ger
    (10) tbmv (just merged this PR)
    (11) trsv
    (12) trsm
    (13) trtri
    (14) gemm

B) Each .cpp have a corresponding .hpp file with it's own template (rocblas_func_(batched)_template) which can call another shared function (ex. reduction). This is currently used in:
    (1) asum
    (2) axpy
    (3) iamax
    (4) iamin
    (5) nrm2
    (6) syr
    (7) trmv

I don't really have a horse in this race myself, but I do think that we should choose one sooner rather than later so we don't have to make even more changes later. Either way, I think we should consult the team in the new year.

Reply:
I agree with you, I think for the time being either ways are fine. IMHO, it sometimes means a lot of 'parameters' and make the code less readable. But I have the disadvantage that B/ is half  shorter than A/.  I will thread on it.

Reply:
This topic is a good candidate for a rough design pattern specification as a wiki page.  Syr I started in September before the load_batch was standardized, I can refactor if requested to fit A style.   Need Juan's input as what he wants for reuse, if consistent that will be much simpler.   But optimizations may also come into play...

Reply:
I am going to ask Juan's feedback